### PR TITLE
[RFC] rust-embedded teams and maintenance of the embedded ecosystem

### LINKED ISSUE #46: Maintenance of core parts of the embedded / Cortex-M ecosystem
A.K.A. @japaric is a bottleneck for development :-)

There are critical components of the embedded / Cortex-M ecosystem that are currently being
maintained by me. Unfortunately I don't have the bandwidth to keep up with everything so issues and
PRs have been piling up in some repos.

This issue is a call for maintainers (if you want to help leave a comment!) and a place to discuss
how to organize maintainership duties.

First, these are the crates that should be maintained by more people:

- `embedded-hal`

Abstractions for things like Serial, SPI, I2C, etc. to build generic drivers that work with AVR, ARM
Cortex-M, MSP430, RISCV, etc. microcontrollers.

There's a bunch to do here as some abstractions are missing and async support is currently
nonexistent. Most of the work to do here is ensuring that the established [process] for adding new
traits is followed and that discussions move towards consensus.

[process]: https://github.com/japaric/embedded-hal#how-to-add-a-new-trait

Quite a few people have volunteered to help and are now collaborators so I think this one is well
covered for now.

- `svd2rust`

Tool used to transform [CMSIS SVD] files (XML files that describe the hardware registers of a
device) into code for low level manipulation of hardware registers. Although the SVD spec was
created by ARM `svd2rust` is being used for both ARM Cortex-M and MSP430 devices. In the later case
TI provides DSLite files (their version of SVD files) and those are converted to SVD files using the
dslite2svd tool.

[CMSIS SVD]: https://github.com/posborne/cmsis-svd

It's expected that `svd2rust` will also grow AVR support -- Atmel has its own description file
format: atdf -- using the atdf2svd approach. And there are XSVD (JSON version of SVD files) [files]
for RISCV -- I don't think they are official though? -- so `svd2rust` could also grow support for
RISCV via a xsvd2svd tool.

[files]: https://github.com/micro-os-plus/sifive-devices-xpack/tree/xpack/xsvd

Unfortunately `svd2rust` has grown into an unmaintainable mess. The test suite takes a lot of time
to run (it takes longer than the CI limit) and it only checks if the generated code still compiles
-- it provides no feedback about whether the generated code exposes a certain interface or not. The
underlying parser lacks error handling and simply panics on errors. Also, `svd2rust` doesn't give
you helpful error messages when your SVD file has some problem -- this is an important feature to
have when manually editing SVD files.

So both `svd2rust` and the underlying `svd` parser need to be refactor (read: rewritten) to make
them more modular and testable. I think the approach to take here is a refactor like the one
proposed in japaric/svd#37. Basically we should be able to test code generation on *parts* of a SVD
file. For example that `<enumeratedValues>..</enumeratedValues>` produces this `enum Foo { .. }`
and this API `impl Foo { fn is_variant(&self) -> bool { .. } }`, etc. This will let us write unit
tests that can replace the huge existing tests that only check for compilation errors. And while you
are at it you can add error handling to the parser and nice error messages to `svd2rust`.

This is a huge task but it will make it easier to continue the development of `svd2rust`. There a
bunch of open issues related to SVD files not being supported; the refactor should make it easier to
fix those issues.

If you want to tackle this let me know in the comments!

---

Then we have core `cortex-m` crates.

- [`cortex-m`]

[`cortex-m`]: https://github.com/japaric/cortex-m

This repo provides access to Cortex-M specific features like access to registers, safe wrappers around
assembly instructions and access to core peripherals.

One old issue in this repo is providing a `svd2rust` generated API to access the core peripherals.
The current API has been manually written and is not as type safe as a `svd2rust` one would be.
Fixing this requires writing or putting together an SVD file that contains only core peripherals,
running `svd2rust` over it and committing the code to the `cortex-m` repo.

- [`cortex-m-rt`]

[`cortex-m-rt`]: https://github.com/japaric/cortex-m-rt

This crate takes care of communicating the memory layout of a program to the linker and providing
implementations of language items that are required to build a `no_std` program.

There are a few missing features in this crate: support for placing functions / interrupt handlers /
the whole program in RAM, relocation of the vector table, etc. that need a design. It may be useful
to coordinate with the representatives of the other embedded targets to see if we can provide a
similar user interface for these features.

- [`cortex-m-quickstart`]

[`cortex-m-quickstart`]: https://github.com/japaric/cortex-m-quickstart

This crate is used as a Cargo project template for writing Cortex-M applications.

Every now and then a user figures a new way to misfollow the instructions and encounters a new error
message. These errors should be documented in the [troubleshooting guide].

[troubleshooting guide]: https://docs.rs/cortex-m-quickstart/0.2.4/cortex_m_quickstart/#troubleshooting

---

Regarding organization I think it may make sense to move these core components under the
[rust-embedded] org and then create, say, a cortex-m team in charge of maintaining the cortex-m
crates. As the AVR, MSP430 and RISCV communities grow they may also want to move core components
into the rust-embedded org and create their own maintainer teams. Thoughts on this?

[rust-embedded]: https://github.com/rust-embedded

cc @ryankurte @Emilgardis *might* be interested in picking up svd2rust development
cc @dylanmckay (AVR) @pftbest (MSP430) @dvc94ch (RISCV) see last paragraph, also `svd2rust` support

### PR DESCRIPTION
**Pull request by japaric**
closes #46 

[Rendered](https://github.com/rust-lang-nursery/embedded-wg/blob/teams/rfcs/0000-teams.md).

RFCs for creating each team

- MSP430 team #137
- RISCV team #138 
- embedded Linux team #139 
- resources team #140 
- tools team #141 
- HAL team #142 
- Cortex-M team #143 
- triage team #144

### CODE REVIEW

**Comment by korken89** [comment]
Hi, I think `painic_abort` is missing in `cortex-m team`, or maybe it is excluded for a reason?

**Comment by japaric** [comment]
@korken89 `panic_abort` is not Cortex-M specific. It also doesn't compile on stable because of `intrinsics::abort`, which doesn't have a path for stabilization.

OTOH, `panic-itm` is in there because ITM is Cortex-M specific, `panic-semihosting` is in there because it only supports the Cortex(-M) at the moment.

**Comment by japaric** [comment]
The RFC has been updated with the input from [today's WG meeting](https://github.com/rust-lang-nursery/embedded-wg/blob/master/minutes/2018-07-30.md). The section on the initial composition of the teams has been removed -- in its place we'll have one RFC PR per team -- to reduce the volume of comments in this thread.

cc @texitoi @korken89 @levex @ithinuel @kellerkindt

**Comment by jamesmunns** [comment]
@japaric one item I didn't see explicitly mentioned is who makes up the `rust-embedded/wg` team. Do we retain the [current roster], would this be reduced to just you, or made up of the leads of each of the sub team/projects?

It would also be good to state how the `wg` team fits in to the other projects in the new organization.

[current roster]: https://github.com/rust-lang-nursery/embedded-wg/tree/teams#roster

**Comment by jamesmunns** [comment]
Rough status of the vote so far:

- [ ] @dvc94ch 
- [ ] @dylanmckay 
- [x] @hannobraun 
- [x] @jamesmunns
- [x] @japaric 
- [ ] @jcsoo 
- [ ] @pftbest 
- [ ] @thejpster

**Comment by japaric** [comment]
I found that the triage procedure used in rust-lang/rust is documented in [the forge](https://forge.rust-lang.org/triage-procedure.html) so I have updated to RFC to replace my guesstimated instructions with a link to the instructions in the forge.

**Comment by ryankurte** [comment]
Are we going to have a separate team for infrastructure / ops (looking after domains/DNS/websites/blogs/etc.) or would that fall within the resources team?

**Comment by jamesmunns** [comment]
Hey @ryankurte, I think that would fall to the resources group, though on https://github.com/rust-lang-nursery/embedded-wg/issues/148 I am pushing for as few self-managed options as possible.

This would leave ownership pretty much with the contributors to the individual repos in the organization (for content and CI).

@nastevens currently controls the DNS for the rust-embedded domain, so we would need to work with him on how to manage this.

**Comment by japaric** [comment]
@ryankurte  We discussed this on today's meeting and decided we'll revisit the creation of an ops team after we have a decision on RFC #148 (Community plan).

As there are no unresolved concerns here, or leftover from the meetings, and there's voting majority I'm going to merge this RFC and make it effective. Thanks everyone for your input here and on IRC.

**Comment by japaric** [comment]
One more comment: all team members now have "bors rights" and can send PRs to bors using "bors r+" (not "@bors r+"; we are using https://bors.tech, not the rust-lang instance). You can find documentation about the bors commands in https://bors.tech/documentation/ 

cc @rust-embedded/all

**Comment by japaric** [comment]
Another update: all team members should now have publish / yank permissions on crates.io. One thing I didn't realize until it was too late is that team members can *not* add or remove owners. I realized this after I removed myself as an owner on most Cortex-M crates so now all those crates are owned by a single team and no more owners can be added. Which is not too bad except that, for example, the itm crate only owner is the cortex-m team but it was supposed that the tools team would also be an owner of that crate -- that can't be done anymore: the cortex-m team members can't add the tools team as an owner.

If this turns out to be a problem in the future we can always ~bribe~ ask someone on the crates.io team to add a user owner to these crates where the only owner is a team.

TL;DR @pftbest @dvc94ch on crates.io add team owners to your crates but don't remove yourselves from the owner lists. Always keep a user owner so they can modify the owner list in the future.
